### PR TITLE
Test REST API with the Docker test framework

### LINKED ISSUE #722: Add tests for the REST API in the docker test framework
- We want to test every single endpoint in the rest API if it works as expected, especially regarding race conditions with other endpoint calls.

### LINKED ISSUE #792: Add status codes map to API pkg
Currently we always return 503 internal server error, even if we should return 404 for "block not found" etc.
We should check all endpoints and return the correct http status codes, as well as the correct RPC codes in the inx server methods.
Maybe it would make sense to add a map that resolves the error to the correct http error code.

### PR DESCRIPTION
**Pull request by daria305**
Corresponding iota.go [PR](https://github.com/iotaledger/iota.go/pull/707)

- add tests for REST API Fixes #722 
- correct http error codes, e.g. use not Found instead of internal server error. Fixes #792 
- add tests for error codes requests

### CODE REVIEW

File: components/restapi/core/component.go [hunk 1]
```diff
@@ -152,7 +152,7 @@ func configure() error {
 		}
 
 		return responseByHeader(c, resp)
-	})
+	}, checkNodeSynced())
```
Comment (muXxer) [inline components/restapi/core/component.go]: Not all endpoints need the node to be synced.
Values that never change after they are created can safely returned by an unsync node.
For example "UTXO changes by slot" are no problem. Either the unsynced node has them or not. No need to be synced. But for transaction metadata or block metadata endpoints, the provided information could be outdated if the node is unsync.

File: pkg/requesthandler/accounts.go [hunk 2]
```diff
@@ -99,7 +99,7 @@ func (r *RequestHandler) ValidatorByAccountAddress(accountAddress *iotago.Accoun
 func (r *RequestHandler) RewardsByOutputID(outputID iotago.OutputID, slot iotago.SlotIndex) (*api.ManaRewardsResponse, error) {
 	utxoOutput, err := r.protocol.Engines.Main.Get().Ledger.Output(outputID)
 	if err != nil {
-		return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to get output %s from ledger: %s", outputID.ToHex(), err)
+		return nil, ierrors.Wrapf(echo.ErrNotFound, "output %s not found in the Ledger: %s", outputID.ToHex(), err)
```
Comment (muXxer) [inline pkg/requesthandler/accounts.go]: This would return `ErrNotFound` in every possible error case.
It should only be `ErrNotFound` if the database entry was not found, otherwise it should be `ErrInternalServerError`.

File: pkg/requesthandler/commitments.go [hunk 3]
```diff
@@ -38,7 +38,7 @@ func (r *RequestHandler) GetCommitmentByID(commitmentID iotago.CommitmentID) (*m
 
 	commitment, err := r.protocol.Engines.Main.Get().Storage.Commitments().Load(commitmentID.Slot())
 	if err != nil {
-		return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to load commitment, commitmentID: %s, slot: %d, error: %w", commitmentID, commitmentID.Slot(), err)
+		return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to load commitment, commitmentID: %s, slot: %d, error: %v", commitmentID, commitmentID.Slot(), err)
```
Comment (muXxer) [inline pkg/requesthandler/commitments.go]: I don't understand the change from `%w` to `%v`.
With that change, the `ierrors` lib is unable to properly wrap and chain the errors.

Can you revert these changes?

In this special case you even could do something like this to be more clean:
```suggestion
		return nil, ierrors.Join(echo.ErrInternalServerError, ierrors.Wrapf(err, "failed to load commitment, commitmentID: %s, slot: %d", commitmentID, commitmentID.Slot()))
```

File: pkg/requesthandler/transaction.go [hunk 4]
```diff
@@ -16,7 +16,7 @@ func (r *RequestHandler) BlockIDFromTransactionID(transactionID iotago.Transacti
 
 	output, spent, err := r.protocol.Engines.Main.Get().Ledger.OutputOrSpent(outputID)
 	if err != nil {
-		return iotago.EmptyBlockID, ierrors.Wrapf(echo.ErrInternalServerError, "failed to get output %s: %s", outputID.ToHex(), err)
+		return iotago.EmptyBlockID, ierrors.Wrapf(echo.ErrNotFound, "output %s of transaction %s not found: %s", transactionID.ToHex(), outputID.ToHex(), err)
```
Comment (muXxer) [inline pkg/requesthandler/transaction.go]: same here

File: pkg/requesthandler/utxo.go [hunk 5]
```diff
@@ -12,7 +12,7 @@ import (
 func (r *RequestHandler) OutputFromOutputID(outputID iotago.OutputID) (*api.OutputResponse, error) {
 	output, err := r.protocol.Engines.Main.Get().Ledger.Output(outputID)
 	if err != nil {
-		return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to get output %s from the Ledger: %s", outputID.ToHex(), err)
+		return nil, ierrors.Wrapf(echo.ErrNotFound, "output %s not found in the Ledger: %s", outputID.ToHex(), err)
```
Comment (muXxer) [inline pkg/requesthandler/utxo.go]: same here

File: pkg/requesthandler/utxo.go [hunk 6]
```diff
@@ -24,7 +24,7 @@ func (r *RequestHandler) OutputFromOutputID(outputID iotago.OutputID) (*api.Outp
 func (r *RequestHandler) OutputMetadataFromOutputID(outputID iotago.OutputID) (*api.OutputMetadata, error) {
 	output, spent, err := r.protocol.Engines.Main.Get().Ledger.OutputOrSpent(outputID)
 	if err != nil {
-		return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to get output %s from the Ledger: %s", outputID.ToHex(), err)
+		return nil, ierrors.Wrapf(echo.ErrNotFound, "output %s not found in the ledger: %s", outputID.ToHex(), err)
```
Comment (muXxer) [inline pkg/requesthandler/utxo.go]: same here

File: pkg/requesthandler/utxo.go [hunk 7]
```diff
@@ -37,7 +37,7 @@ func (r *RequestHandler) OutputMetadataFromOutputID(outputID iotago.OutputID) (*
 func (r *RequestHandler) OutputWithMetadataFromOutputID(outputID iotago.OutputID) (*api.OutputWithMetadataResponse, error) {
 	output, spent, err := r.protocol.Engines.Main.Get().Ledger.OutputOrSpent(outputID)
 	if err != nil {
-		return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to get output %s from the Ledger: %s", outputID.ToHex(), err)
+		return nil, ierrors.Wrapf(echo.ErrNotFound, "output %s not found in the Ledger: %s", outputID.ToHex(), err)
```
Comment (muXxer) [inline pkg/requesthandler/utxo.go]: same here

File: pkg/requesthandler/blocks.go [hunk 8]
```diff
@@ -23,7 +23,7 @@ func (r *RequestHandler) BlockByID(blockID iotago.BlockID) (*iotago.Block, error
 func (r *RequestHandler) BlockMetadataByBlockID(blockID iotago.BlockID) (*api.BlockMetadataResponse, error) {
 	blockMetadata, err := r.protocol.Engines.Main.Get().BlockRetainer.BlockMetadata(blockID)
 	if err != nil {
-		return nil, ierrors.Wrapf(echo.ErrInternalServerError, "failed to get block metadata %s: %s", blockID.ToHex(), err)
+		return nil, ierrors.Wrapf(echo.ErrNotFound, "metadata not found for block %s: %s", blockID.ToHex(), err)
```
Comment (muXxer) [inline pkg/requesthandler/blocks.go]: same here

**Review by muXxer — CHANGES_REQUESTED** [review]

File: components/restapi/core/component.go (hunk 1 above)
Comment (daria305) [reply components/restapi/core/component.go]: I see, I left the ones added for OutputMetadata

File: tools/docker-network/tests/eventapi_test.go [hunk 9]
```diff
@@ -413,11 +413,8 @@ func test_NFTTransactionBlocks(t *testing.T, d *DockerTestFramework) {
 			assertion()
 		}
 
-		// d.AssertTransactionMetadataByTransactionID(ctx, eventClt, outputId.TransactionID(), finish)
-		// d.AssertTransactionMetadataIncludedBlocks(ctx, eventClt, outputId.TransactionID(), finish)
```
Comment (jkrvivian) [inline tools/docker-network/tests/eventapi_test.go:417]: Can you add a TODO comment to this file that reminds us to test transactionMetadata topics?

File: tools/docker-network/tests/dockerframework.go [hunk 10]
```diff
@@ -349,105 +359,136 @@ func (d *DockerTestFramework) StopIssueCandidacyPayload(nodes ...*Node) {
...
 	issuerResp, congestionResp := d.PrepareBlockIssuance(ctx, clt, issuer.Address)
 
-	return d.CreateBlock(ctx, &iotago.TaggedData{
+	return d.CreateBlock(&iotago.TaggedData{
 		Tag: tag,
-	}, issuerId, congestionResp, issuerResp)
+	}, issuerID, congestionResp, issuerResp)
+}
+
+func (d *DockerTestFramework) CreateValueBlock(issuerAccountID iotago.AccountID) (*iotago.Block, *iotago.SignedTransaction, *OutputData, iotago.Output) {
```
Comment (jkrvivian) [inline tools/docker-network/tests/dockerframework.go]: Rename it to `CreateBasicOutputBlock`.

File: tools/docker-network/tests/dockerframework.go [hunk 11]
```diff
@@ -349,105 +359,136 @@ func (d *DockerTestFramework) StopIssueCandidacyPayload(nodes ...*Node) {
...
+		AddInput(&builder.TxInput{
+			UnlockTarget: input.Address,
+			InputID:      input.ID,
+			Input:        input.Output,
+		}).
+		AddOutput(basicOutput).
+		SetCreationSlot(currentSlot).
+		AllotAllMana(currentSlot, issuerAccountID, 0).
+		Build()
+	require.NoError(d.wallet.Testing, err)
```
Comment (jkrvivian) [inline tools/docker-network/tests/dockerframework.go]: Let's create a `CreateBasicOutputFromInput` method in wallet and move the tx building logic there, so it's aligned to other methods.

File: tools/docker-network/tests/coreapi_test.go [hunk 12]
```diff
@@ -0,0 +1,542 @@
...
+				require.Nil(t, respBlock)
+			},
+		},
+		{
+			name: "Test_BlockMetadataByBlockID_Failure",
+			testFunc: func(t *testing.T, nodeAlias string) {
+				blockID := tpkg.RandBlockID()
+				resp, err := d.wallet.Clients[nodeAlias].BlockMetadataByBlockID(context.Background(), blockID)
+				require.Error(t, err)
+				require.True(t, isStatusCode(err, http.StatusInternalServerError))
```
Comment (jkrvivian) [inline tools/docker-network/tests/coreapi_test.go]: This should return NotFound?

File: tools/docker-network/tests/coreapi_test.go [hunk 13]
```diff
@@ -0,0 +1,542 @@
...
+					require.NoError(t, err)
+					commitment, err := d.wallet.Clients[nodeAlias].CommitmentBySlot(context.Background(), infoRes.Status.LatestCommitmentID.Slot())
+					require.NoError(t, err)
+
+					resp, err = d.wallet.Clients[nodeAlias].Congestion(context.Background(), accountAddress, 0, commitment.MustID())
+					require.NoError(t, err)
+					require.NotNil(t, resp)
+					// later we check if all nodes have returned the same BIC value for this account
+					bicPerNoode[nodeAlias] = resp.BlockIssuanceCredits
+				})
```
Comment (jkrvivian) [inline tools/docker-network/tests/coreapi_test.go:277]: We can call assertBICs here, no?

File: components/restapi/core/accounts.go [hunk 14]
```diff
@@ -27,7 +27,7 @@ func congestionByAccountAddress(c echo.Context) (*api.CongestionResponse, error)
 		workScores = append(workScores, workScore)
 	}
 
-	commitment, err := deps.RequestHandler.GetCommitmentByID(commitmentID)
+	commitment, err := deps.RequestHandler.GetCommitmentByID(queryCommitmentID)
```
Comment (jkrvivian) [inline components/restapi/core/accounts.go]: `queryCommitment` then? :grimacing:

File: components/restapi/core/accounts.go [hunk 15]
```diff
@@ -43,6 +43,16 @@ func congestionByAccountAddress(c echo.Context) (*api.CongestionResponse, error)
...
+	maxCommittableAge := deps.RequestHandler.CommittedAPI().ProtocolParameters().MaxCommittableAge()
+	latestCommittedSlot := deps.RequestHandler.GetLatestCommitment().Slot()
+	if latestCommittedSlot >= maxCommittableAge && commitment.Slot()+maxCommittableAge < latestCommittedSlot && queryCommitmentID != iotago.EmptyCommitmentID {
+		return nil, ierrors.Wrapf(echo.ErrBadRequest, "invalid commitmentID, target slot index older than allowed (%d<%d)", commitment.Slot(), latestCommittedSlot-maxCommittableAge)
+	}
+
+	if commitment.Slot() > latestCommittedSlot && queryCommitmentID != iotago.EmptyCommitmentID {
+		return nil, ierrors.Wrapf(echo.ErrBadRequest, "invalid commitmentID, slot %d is not committed yet, latest committed slot: %d", commitment.Slot(), latestCommittedSlot)
+	}
+
```
Comment (jkrvivian) [inline components/restapi/core/accounts.go]: Move this check right after we parse queryCommitmentID, so we can return this error asap.

```go
if queryCommitmentID != iotago.EmptyCommitmentID {
    if commitment.Slot() > latestCommittedSlot {...}

    if latestCommittedSlot >= maxCommittableAge && commitment.Slot()+maxCommittableAge < latestCommittedSlot {...}

}
```

**Review by jkrvivian — CHANGES_REQUESTED** [review]
Thanks for fixing all linter errors!

File: components/restapi/core/accounts.go (hunk 14 above)
Comment (daria305) [reply components/restapi/core/accounts.go]: but that's `commitmentID`

File: tools/docker-network/tests/coreapi_test.go (hunk 13 above)
Comment (daria305) [reply tools/docker-network/tests/coreapi_test.go:277]: This check if all nodes returned the same and iterates over `bicPerNoode map`, so it should be called afterwards

File: tools/docker-network/tests/coreapi_test.go (hunk 12 above)
Comment (daria305) [reply tools/docker-network/tests/coreapi_test.go]: It was like that, but this still can return error for getting the store, so it's better to use InternalServerErr, see the above MuXxer comment

File: components/restapi/core/accounts.go (hunk 14 above)
Comment (jkrvivian) [reply components/restapi/core/accounts.go]: I meant the returned commitment:
```suggestion
	queryCommitment, err := deps.RequestHandler.GetCommitmentByID(queryCommitmentID)
```

**Review by jkrvivian — APPROVED** [review]

**Review by muXxer — APPROVED** [review]
